src/main/kotlin/io/sip3/salto/ce/sdp/SdpHandler.kt

```diff
@@ -0,0 +1,175 @@
+                readCodecs(config)
+            } catch (e: Exception) {
+                logger.error("SdpHandler `readCodecs()` failed.", e)
+            }
+        }
+
+        vertx.eventBus().localConsumer<SIPMessage>(RoutesCE.sdp_session_update) { event ->
+            try {
+                val message = event.body()
+                handleSdp(message)
+            } catch (e: Exception) {
+                logger.error("SdpHandler 'handleSdp()' failed.", e)
+            }
+        }
+
+        vertx.eventBus().localConsumer<String>(RoutesCE.sdp_session_terminate) { event ->
+            try {
+                val sessionKey = event.body()
+                sessions.remove(sessionKey)
+            } catch (e: Exception) {
+                logger.error("SdpHandler sdpSession remove failed.", e)
+            }
+        }
+    }
+
+    private fun readCodecs(config: JsonObject) {
+        val tmpCodecs = mutableMapOf<String, Codec>()
+        config.getJsonArray("codecs")?.forEach { codecObject ->
+            val codec = (codecObject as JsonObject).mapTo(Codec::class.java)
+            tmpCodecs[codec.name!!] = codec
+        }
+        codecs = tmpCodecs
+    }
+
+    private fun handleSdp(message: SIPMessage) {
+        val sessionDescription: SessionDescription = getSessionDescription(message) ?: return
+
+        val callId = message.callId()!!
+        sessions.getOrPut(callId) { SdpSessionDescription(callId) }.apply {
+            when (message) {
+                is SIPRequest -> {
+                    request = sessionDescription
+                }
+                is SIPResponse -> {
+                    response = sessionDescription
+                }
+            }
+
+            if (isFilled()) {
+                sendSdpSession(this)
+            }
+        }
+    }
+
+    private fun sendSdpSession(sdpSessionDescription: SdpSessionDescription) {
+        val requestMediaDescription = sdpSessionDescription.request.getMediaDescription("audio")
+        val responseMediaDescription = sdpSessionDescription.response.getMediaDescription("audio")
+
+        val callId = sdpSessionDescription.callId
+        val codec = getCodec(requestMediaDescription, responseMediaDescription)
+        val requestSdpSession = createSdpSession(callId, requestMediaDescription, codec)
+        val responseSdpSession = createSdpSession(callId, responseMediaDescription, codec)
+
+        // Send to Management Socket
+        vertx.eventBus().send(RoutesCE.sdp_session, listOf(requestSdpSession, responseSdpSession), USE_LOCAL_CODEC)
+    }
+
+    private fun getCodec(request: MediaDescriptionField, response: MediaDescriptionField): Codec {
+        val payloadTypes = request.payloadTypes.intersect(response.payloadTypes.asIterable())
+        val payloadType = payloadTypes.first()
+        val payload: RtpMapAttribute? = response.getFormat(payloadType)
+                ?: request.getFormat(payloadType)
+
+        val codec = codecs[payload?.codec]
```

Comment:
Let's build codecs map by payloadType? Seems that it will be way more efficient

Reply:
One codec may have range of payload types, e.g. `GSM-EFR`. That's why payloadType can't be used as codecs map key.